### ISSUE TITLE
Users want to be able to do http -> https redirection with XFP

### ISSUE THREAD

**Issue by nbkrause** [datawire/ambassador#1233]
Pre Ambassador 0.50.0, setting `x_forwarded_proto_redirect: true` in the Ambassador `Module` configured Ambassador to read XFP for and redirect if it was HTTP. This is useful when terminating TLS at an L7 LB.

**Comment by bbl232** [comment]
This is exactly our use case. GCLB (terminating TLS) -> K8S cluster hitting our nodeport service for ambassador

**Comment by justinfiore** [comment]
We have the same issue (using AWS Layer 7 LBs).

**Comment by adrianchifor** [comment]
This has stopped working for us as well in 0.50+

**Comment by richarddli** [comment]
@justinfiore @bbl232 @adrianchifor Have you guys tried using `redirect_cleartext_from`?

Digging through the past issues, @concaf had a fix for this in https://github.com/datawire/ambassador/pull/1193 but it enabled XFP by breaking XFF, which isn't desirable.

So the right fix is to enable XFP *without* breaking XFF. But the first step is to confirm that `redirect_cleartext_from` doesn't work (and why).

**Comment by bbl232** [comment]
I had tried this yeah as per my slack thread: https://datawire-oss.slack.com/archives/CAULN7S76/p1549404948562400

**Comment by adrianchifor** [comment]
`redirect_cleartext_from` doesn't work for us either.

**Comment by gsagula** [comment]
@adrianchifor @bbl232 Are using ` x_forwarded_proto_redirect: true` and `use_remote_addess: true`? If yes, I believe that this logic is causing the problem:
https://github.com/envoyproxy/envoy/blob/b771f99f53142bbb1496dbcddccc5defaaaa8dd8/source/common/http/conn_manager_utility.cc#L108

Assuming that `x_forwarded_proto_redirect` relies on `x-forwarded-proto: https` header to do the redirection, this will never happen since the connection with the external LB is no secure. In this case, whatever XFP header sent from the LB will be replaced by `x-forwarded-proto: http`.

**Comment by bbl232** [comment]
I didn't have anything set for `use_remote_address` so that would just be whatever the default is

**Comment by adrianchifor** [comment]
Don't have `use_remote_address` either. Ambassador is terminating 443 TLS from the LB. `x_forwarded_proto_redirect: true` worked as expected before 0.50.

**Comment by nbkrause** [comment]
@gsagula @bbl232 @adrianchifor `use_remote_address` defaults to `true` since `0.50.0`

Does explicitly setting that to `false` in the ambassador `Module` change the behavior at all? 

**Note:** you will need to restart Ambassador for the `use_remote_address` configuration change to take effect

**Comment by adrianchifor** [comment]
Will try first thing tomorrow, thanks.

**Comment by adrianchifor** [comment]
@nbkrause No luck
```
apiVersion: "ambassador/v1"
kind: "Module"
name: "ambassador"
config:
  x_forwarded_proto_redirect: true
  use_remote_address: false
```
Set this and restarted ambassador pods, still not redirecting to https.

**Comment by gsagula** [comment]
Thanks for checking it. I will look into the issue again later.

**Comment by gsagula** [comment]
@bbl232 Do you have an example config that works with `0.40.2`?

**Comment by bbl232** [comment]
```yaml
  service:
    type: NodePort
    httpNodePort: 30000
    annotations:
      getambassador.io/config: |
        ---
        apiVersion: ambassador/v0
        kind: TracingService
        name: tracing
        service: "zipkin.istio-system:9411"
        driver: zipkin
        config: {}
        tag_headers:
          - ":authority"
          - ":path"
        ---
        apiVersion: ambassador/v0
        kind: Module
        name: ambassador
        config:
          x_forwarded_proto_redirect: true
```

**Comment by gsagula** [comment]
The problem seems to be Ambassador that is not setting the `require_tls: ALL` in Envoy route_config.
```
route_config:
            name: local_route
            virtual_hosts:
            - name: echo-service
              require_tls: ALL
```

**Comment by gsagula** [comment]
@bbl232 @adrianchifor I believe that this is fixed now, but note that since Ambassador was designed to be an edge proxy, `use_remote_address` defaults to `true`. You need to set it to `false` for ` x_forwarded_proto_redirect` to work. Please, let me know if you have any question.

**Comment by kflynn** [comment]
Fixed in 0.60.0
